File: packages/shared/src/types/pagination.ts
/**
 * Shared pagination envelope for list endpoints.
 */
export interface PaginatedResult<T> {
  items: T[];
  page: number;
  limit: number;
  total: number;
  totalPages: number;
}

export const MAX_PAGE_SIZE = 100;
export const DEFAULT_PAGE_SIZE = 25;

File: apps/backend/schemas/pagination.py
"""Generic paginated-list envelope — mirrors packages/shared's
PaginatedResult<T> and its DEFAULT_PAGE_SIZE/MAX_PAGE_SIZE constants."""

from __future__ import annotations

from pydantic import BaseModel

DEFAULT_PAGE_SIZE = 25
MAX_PAGE_SIZE = 100


class PaginatedResult[T](BaseModel):
    items: list[T]
    page: int
    limit: int
    total: int
    totalPages: int


def safe_page(page: int) -> int:
    return max(1, page)


def safe_limit(limit: int) -> int:
    return min(MAX_PAGE_SIZE, max(1, limit))


def total_pages(total: int, limit: int) -> int:
    return max(1, -(-total // limit))  # ceil division without importing math
